tests/model_serving/model_server/authentication/utils.py

```diff
@@ -0,0 +1,111 @@
+import base64
+import re
+from contextlib import contextmanager
+from typing import Dict, List, Optional
+
+from kubernetes.dynamic import DynamicClient
+from ocp_resources.inference_service import InferenceService
+from ocp_resources.role import Role
+from simple_logger.logger import get_logger
+
+from utilities.inference_utils import Inference
+
+LOGGER = get_logger(name=__name__)
+
+
+def base64_encode_str(text: str) -> str:
+    return base64.b64encode(text.encode()).decode()
+
+
+def verify_inference_response(
+    inference_service: InferenceService,
+    runtime: str,
+    inference_type: str,
+    protocol: str,
+    model_name: str,
+    inference_text: str,
+    expected_response_text: Optional[str] = "",
+    insecure: bool = True,
+    token: Optional[str] = None,
+    authorized_user: Optional[bool] = None,
+) -> None:
+    inference = Inference(
+        inference_service=inference_service,
+        runtime=runtime,
+        inference_type=inference_type,
+        protocol=protocol,
+    )
+
+    res = inference.run_inference(
+        model_name=model_name,
+        text=inference_text,
+        token=token,
+        insecure=insecure,
+    )
+
+    if authorized_user is False:
+        auth_header = "x-ext-auth-reason"
+
+        if auth_reason := re.search(rf"{auth_header}: (.*)", res["output"], re.MULTILINE):
+            reason = auth_reason.group(1).lower()
+
+            if token:
+                assert re.search(r"not (?:authenticated|authorized)", reason)
+
+            else:
+                assert "credential not found" in reason
+
+        else:
+            LOGGER.error(f"Auth header {auth_header} not found in response. Response: {res['output']}")
+            raise
+
+    elif inference.inference_response_text_key_name:
+        assert res["output"][inference.inference_response_text_key_name] == expected_response_text
+
+    elif output := re.findall(r"generated_text\": \"(.*)\"", res["output"], re.MULTILINE):
+        assert "".join(output) == expected_response_text
```

Comment:
This is not a great way to define the validation of inference outputs, it is rare that you get the exact same response every single time

Reply:
please check `INFERENCE_QUERIES` not sure how this can change?

Reply:
In ods-ci we have a few different approaches to handle potentially different responses from models. 
In the case of flan-t5-small I think we are ok with a `==` check, but there are other models where the response will be slightly different every time (still text based) and we use a similarity score between expected and received text. 
There are also models that do not respond with text but rather with images or arrays of floats (that can vary slightly due to imperfect rounding), and in both cases there we had to define helper functions to parse the results and handle validation. 
There is no one size fits all, but perhaps it could make sense to start setting up this function to call other helper functions mapped to the model being used, rather than having the check here directly.

Reply:
Do we even want to verify the actual content of the response? We are not testing the model itself, so if a given model answers nonsense, it's a model issue, not an ODH issue IMO. We could try to assert that the content of the answer makes sense, but I think that is a bit out of scope for us, and can potentially make the test flaky due to the reasons highlighted by @lugi0 . I think that in this case just verifying that the answer contains some non-empty string should be enough.